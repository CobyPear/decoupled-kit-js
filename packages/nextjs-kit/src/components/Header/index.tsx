import {
	Header as ReactKitHeader,
	isNavItemArray,
	type NavHeaderProps,
	type NavItem,
} from '@pantheon-systems/react-kit/components/Header';
import Link from 'next/link';

const Header = (props: NavHeaderProps) => {
	return (
		<ReactKitHeader
			{...props}
			linkComponent={props?.linkComponent ? props.linkComponent : Link}
		/>
	);
};

export { Header, NavHeaderProps, NavItem, isNavItemArray };
