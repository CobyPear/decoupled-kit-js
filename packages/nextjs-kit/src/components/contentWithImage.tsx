import React from 'react';
import Image from 'next/image';
import Link from 'next/link';
import type { ImageProps } from 'next/image';

interface ContentProps {
	title: string;
	imageProps?: ImageProps;
	content: string;
	date?: string;
	previousPagePath?: string;
}

/**
 *
 * @param props - The props needed for the ContentWithImage component
 * @param props.title - The title of your content
 * @param props.imageProps - All props the user wishes to pass to the next/image component
 * @remarks
 * imageProps is an optional prop to be used if there is an image to be associated with the content.
 * If imageProps is used it is required that the user passes in values for src, width, and height
 * See the documentation link below for more information on optional and required props
 * @see {@link https://nextjs.org/docs/api-reference/next/image} for all next/image documentation
 * @param props.content - A string of elements to make up the content on the post
 * @param props.date - An optional date to be displayed on the post
 * @param props.previousPagePath - The path of the previous page to navigate back to
 * @returns A component with a featured image and content passed by the user
 */
const ContentWithImage: React.FC<ContentProps> = ({
	title,
	imageProps,
	content,
	date,
	previousPagePath,
}: ContentProps) => {
	return (
		<article className="prose lg:prose-xl mt-10 mx-auto">
			<h1>{title}</h1>
			{date ? (
				<p className="text-sm text-gray-600">{new Date(date).toDateString()}</p>
			) : null}

			{previousPagePath ? (
				<Link passHref href={previousPagePath}>
					<a className="font-normal">Back &rarr;</a>
				</Link>
			) : null}

			<div className="mt-12 max-w-lg mx-auto lg:grid-cols-3 lg:max-w-screen-lg">
				{imageProps ? (
					<div
						className="relative w-full rounded-lg shadow-lg overflow-hidden mb-10"
						style={{ height: '50vh' }}
					>
						<Image {...imageProps} />
					</div>
				) : null}
			</div>

			<div
				className="break-words mt-12"
				dangerouslySetInnerHTML={{ __html: content }}
			/>
		</article>
	);
};

export default ContentWithImage;
