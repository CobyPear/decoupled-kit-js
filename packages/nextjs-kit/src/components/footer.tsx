import React from 'react';
import Link from 'next/link';
import { FooterMenuItem } from '../types/index';

interface FooterMenuProps {
  footerMenuItems: FooterMenuItem[];
  children: JSX.Element;
}

/**
 * This is a Footer component.
 *
 * @param {Props} footerMenuItems holds the linkText, href, and parent of each route to be displayed
 * @param {Props} children  copyright display and information to be passed in by the user
 * @returns {React.JSX.Element}
 */
const Footer: React.FC<FooterMenuProps> = ({
  footerMenuItems,
  children,
}: FooterMenuProps) => {
  const FooterMenu = () => {
    const menuArr = [];
    if (footerMenuItems) {
      // some not so great code to account for nested menu elements
      for (let i = 0; i < footerMenuItems.length; i++) {
        if (footerMenuItems[i + 1] && footerMenuItems[i + 1].parent) {
          menuArr.push(
            <ul key={i}>
              <li className="list-disc text-blue-300 ml-3">
                <Link href={footerMenuItems[i].href}>
                  <a className="text-blue-300 hover:underline hover:text-blue-100 focus:text-purple-600 active:text-purple-300">
                    {footerMenuItems[i].linkText}
                  </a>
                </Link>
              </li>
              <li className="list-disc text-blue-300 ml-8">
                <Link href={footerMenuItems[i + 1].href}>
                  <a className="text-blue-300 hover:underline hover:text-blue-100 focus:text-purple-600 active:text-purple-300">
                    {footerMenuItems[i + 1].linkText}
                  </a>
                </Link>
              </li>
            </ul>
          );
          // increment iterator to skip the next render
          i++;
        } else {
          menuArr.push(
            <li key={i} className="list-disc text-blue-300 ml-3">
              <Link href={footerMenuItems[i].href}>
                <a className="text-blue-300 hover:underline hover:text-blue-100 focus:text-purple-600 active:text-purple-300">
                  {footerMenuItems[i].linkText}
                </a>
              </Link>
            </li>
          );
        }
      }
    }
    return (
      <nav className="flex flex-col max-w-lg mx-auto lg:max-w-screen-lg">
        <ul>{menuArr?.map(menu => menu)}</ul>
      </nav>
    );
  };
  return (
    <footer className="w-full text-white bg-black p-4 mt-12">
      <FooterMenu />
      <div className="flex my-4 p-2">{children}</div>
    </footer>
  );
};

export default Footer;
