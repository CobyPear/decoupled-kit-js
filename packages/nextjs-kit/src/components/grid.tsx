import React from 'react';

/**
 *
 * @param cols - The number of columns to build the grid with
 * @param children - The JSX elements used as the content of the grid
 * @returns A style and positioning helper grid component meant to be used with the withGrid component
 * @remarks
 */

export const Grid = ({
	cols = '3',
	children,
}: {
	cols?: string;
	children?: JSX.Element | JSX.Element[];
}) => {
	return (
		<div
			className={`mt-12 grid gap-5 max-w-content mx-auto lg:grid-cols-${String(
				cols,
			)} lg:max-w-screen-lg`}
		>
			{children}
		</div>
	);
};

/**
 *
 * @param props - props to spread on
 * @param children - The JSX elements used as the content of the grid
 * @returns A style and positioning helper grid component meant to be used with the withGrid component
 * @remarks
 */
// a Higher Order Component to use Grid
export const withGrid = (Component: React.ElementType) => {
	const GridedComponent = ({
		data,
		...props
	}: {
		data: Record<string, string | number>[];
		contentType: string;
	}): JSX.Element => {
		return (
			<>
				{data ? (
					<Grid>
						{data.map((content, i) => {
							return <Component key={i} content={content} {...props} />;
						})}
					</Grid>
				) : props.contentType ? (
					<h2 className="text-xl text-center mt-14">
						No {props.contentType} found 🏜
					</h2>
				) : null}
			</>
		);
	};

	return GridedComponent;
};
