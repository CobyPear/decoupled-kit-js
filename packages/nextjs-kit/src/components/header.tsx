import React from 'react';
import Link from 'next/link';
import { HeaderData } from '../types/index';

interface HeaderProps {
  navItems: HeaderData[];
}

/**
 * This is a Header component.
 *
 * @param {HeaderProps} headers holds the title and href of each route to be displayed
 * @returns {React.JSX.Element}
 */

const Header: React.FC<HeaderProps> = ({
  navItems,
}: HeaderProps): JSX.Element => {
  return (
    <div className="my-0 pt-10 px-5 text-xl">
      <nav>
        <ul className="flex flex-row flex-wrap sm:flex-nowrap list-none justify-between max-w-screen-sm mx-auto">
          {navItems.map(item => {
            item.align = item.align !== undefined ? item.align : 'right';
            return (
              <li
                className={`${item.align === 'right' ? 'mr-auto' : 'mx-4'}`}
                key={item.href}
              >
                <Link className="font-sans" href={item.href}>
                  <a className="hover:underline">{item.title}</a>
                </Link>
              </li>
            );
          })}
        </ul>
      </nav>
    </div>
  );
};

export default Header;
