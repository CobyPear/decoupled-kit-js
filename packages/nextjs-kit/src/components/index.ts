export { default as Header } from './header';
export { default as Footer } from './footer';
export { default as Paginator } from './paginator';
export { default as ContentWithImage } from './contentWithImage';
export { default as Recipe } from './recipe';
