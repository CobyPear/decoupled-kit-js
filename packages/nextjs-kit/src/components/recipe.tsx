import React from 'react';
import Image from 'next/image';
import type { ImageProps } from 'next/image';
import { useRouter } from 'next/router';

interface RecipeProps {
	title: string;
	category: string;
	imageProps?: ImageProps;
	ingredients: string[];
	instructions: string;
	children?: JSX.Element;
}

/**
 *
 * @param props - The props needed for the Recipe component
 * @param props.title - The title of the recipe
 * @param props.category - The identifying category of your recipe
 * @param props.imageProps - All props the user wishes to pass to the next/image component
 * @remarks
 * imageProps is an optional prop to be used if there is an image to be associated with the content.
 * If imageProps is used it is required that the user passes in values for src, width, and height
 * See the documentation link below for more information on optional and required props
 * @see {@link https://nextjs.org/docs/api-reference/next/image} for all next/image documentation
 * @param props.ingredients - An array of strings holding ingredients to be displayed
 * @param props.instructions - Instructions from a CMS, usually a string of raw HTML. This string will be set as `dangerouslySetInnerHTML`
 * @param props.previousPagePath - The path of the previous page to navigate back to
 * @returns A recipe component with content and an optional image passed by the user
 */
const Recipe: React.FC<RecipeProps> = ({
	title,
	category,
	imageProps,
	ingredients,
	instructions,
	children,
}: RecipeProps) => {
	const router = useRouter();
	return (
		<article className="prose lg:prose-xl mt-10 mx-auto h-fit p-4 sm:p-0">
			<header>
				<h1>{title}</h1>
				<div className="flex flex-row justify-between">
					<a
						onClick={() => router.back()}
						className="font-normal cursor-pointer"
					>
						Back &rarr;
					</a>
					<span className="text pb-2 pr-3 text-sm text-slate-400">
						{category}
					</span>
				</div>
			</header>
			{imageProps ? (
				<div className="relative max-w-lg mx-auto min-w-full h-[50vh] rounded-lg shadow-lg overflow-hidden mt-12 mb-10">
					<Image {...imageProps} />
				</div>
			) : null}

			<div className="flex flex-col sm:flex-row">
				<section className="flex flex-col min-w-fit sm:border-r-2 pr-4">
					<h2>Ingredients</h2>
					<ul>
						{ingredients?.map((ingredient, i) => {
							if (ingredient.startsWith('For')) {
								return (
									<li className="list-none" key={i}>
										<strong>{ingredient}</strong>
									</li>
								);
							} else {
								return <li key={i}>{ingredient}</li>;
							}
						})}
					</ul>
				</section>
				<section className="flex flex-col pl-4">
					<h2 className="ml-4">Directions</h2>
					<div
						dangerouslySetInnerHTML={{
							__html: instructions,
						}}
					/>
				</section>
			</div>
			{children}
		</article>
	);
};

export default Recipe;
