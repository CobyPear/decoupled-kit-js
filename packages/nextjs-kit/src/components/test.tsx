// This file can be removed when development
// starts in the components directory.
import React from 'react';

interface Props {
  title: string;
  greeting: string;
}

/**
 * This is a test component. This is an example typedoc comment
 *
 * @param {Props} props
 * @returns React.JSX.Element
 */
const Test: React.FC<Props> = (props: Props) => {
  return (
    <>
      <h2>{props.title}</h2>
      <h3>{props.greeting}</h3>
    </>
  );
};

export default Test;
