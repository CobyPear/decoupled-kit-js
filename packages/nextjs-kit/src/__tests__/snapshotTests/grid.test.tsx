import { render } from '@testing-library/react';

import { withGrid } from '../../components/grid';
import exampleGridData from '../data/exampleGridData.json';

/**
 * @vitest-environment jsdom
 */

interface GridifyProps {
	content: {
		title: string;
		body: string;
	};
}

const Gridify: React.FC<GridifyProps> = ({
	content,
}: GridifyProps): JSX.Element => {
	return (
		<div>
			<h1>{content.title}</h1>
			<p>{content.body}</p>
		</div>
	);
};
const GridifyGrid = withGrid(Gridify);

describe('<Grid />', () => {
	it("should render 'Grid'", () => {
		const { asFragment } = render(
			<div>
				<GridifyGrid data={exampleGridData} contentType="grid data" />,
			</div>,
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
