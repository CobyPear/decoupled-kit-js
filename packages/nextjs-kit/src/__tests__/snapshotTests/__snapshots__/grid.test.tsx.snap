// Vitest Snapshot v1

exports[`<Grid /> > should render 'Grid' 1`] = `
<DocumentFragment>
  <div
    class="mt-12 grid gap-5 max-w-content mx-auto lg:max-w-screen-lg lg:grid-cols-3"
  >
    <div>
      <h1>
        Item 1
      </h1>
      <p>
        First item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 2
      </h1>
      <p>
        Second item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 3
      </h1>
      <p>
        Third item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 4
      </h1>
      <p>
        Fourth item in the grid
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Grid /> > should render FallbackComponent instead of 'Grid' 1`] = `
<DocumentFragment>
  <div
    class="mt-12 grid gap-5 max-w-content mx-auto lg:max-w-screen-lg lg:grid-cols-3"
  >
    <div>
      <h1>
        Item 1
      </h1>
      <p>
        First item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 2
      </h1>
      <p>
        Second item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 3
      </h1>
      <p>
        Third item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 4
      </h1>
      <p>
        Fourth item in the grid
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Grid /> > should render a 'Grid' with 4 columns 1`] = `
<DocumentFragment>
  <div
    class="mt-12 grid gap-5 max-w-content mx-auto lg:max-w-screen-lg lg:grid-cols-4"
  >
    <div>
      <h1>
        Item 1
      </h1>
      <p>
        First item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 2
      </h1>
      <p>
        Second item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 3
      </h1>
      <p>
        Third item in the grid
      </p>
    </div>
    <div>
      <h1>
        Item 4
      </h1>
      <p>
        Fourth item in the grid
      </p>
    </div>
  </div>
</DocumentFragment>
`;
